62_CAD/fragment_shader.hlsl

```diff
@@ -372,6 +372,26 @@ float32_t4 calculateFinalColor<true>(const uint2 fragCoord, const float localAlp
     return col;
 }
 
+// TODO: place this MSDF code elsewhere
+float median(float r, float g, float b) {
+    return max(min(r, g), min(max(r, g), b));
+}
+
+float screenPxRange(float2 uv) {
```

Comment:
It's basically this:

`On screen size of the quad` / `MSDF Resolution` * `MSDF Pixel Range`

(From msdfgen docs: `screenPxRange()` represents the distance field range in output screen pixels. For example, if the pixel range was set to 2 when generating a 32x32 distance field, and it is used to draw a quad that is 72x72 pixels on the screen, it should return 4.5 (because 72/32 * 2 = 4.5). For 2D rendering, this can generally be replaced by a precomputed uniform value.)

I did it using the 3D path as we have to support transformations on the text (due to DWG and such), so this allows it to be more flexible (and only costs one fwidth)

Reply:
I think it should be / float2(32, 32); or the msdf texture size instead of `float2(globals.resolution);`
let's double check the math, so we're trying to get from

1. from [0, 1] of msdf sampled value to [-0.5, 0.5] by `(sd - 0.5)`

2. then we multiply by `MsdfPixelRange` to get it to [-MsdfPixelRange/2, MsdfPixelRange/2] and that is basically the 
distance in glyph texture units (because the spectrum/range from lowest to highest sdf is MsdfPixelRange! the input to msdfgen!)
now we need this distance in uv units:

3. we divde by (32,32) or `GetDimensions(msdfTexture)` to get the distance in uv units
now how to turn uv distance into actual screenspace distance?! that's the main question!

here it uses fwidth to do that because it shows you the rate of change of both U and V with regards to x and y on the screen and the inverse of that gives you the screenspace size of the
so if everything is aligned (u with x and v with y) it becomes `1.0 / abs(float2(ddx(u),ddy(v)))`

`1/ddx(u)` gives you the horizontal size of the glyph box in screen pixels!
and `1/ddy(v)` gives you the vertical size of the glyph box in screen pixels!
If you're not sure why read the thought experiment below:

++++
Optional Read 
ddx(u) tells you the rate of change of u per pixel in x direction
if you think about it ddx(u) is basicaly `dU/dScreenPixel` (if you go 1 pixel how much does u value change?)
and 1/ddx(u) is `dScreenPixel/dU` and gives you the inverse rate of change (if you go by 1 units in `u` how many units in screen pixels? ---> width pixels!) 
++++

But fwidth is just used as an approximation. `float2 fwidth = abs(ddx(uv) + ddy(uv))`
small note: ddx(uv) is a float2, it's computing ddx for both u and v.

Now that you understand this, you get why is the `/ float2(globals.resolution)` is buggy and why you need to have different functions for this for 1. glyph boxes 2. hatch fills
and if you fix this you are going to see these bugs fixed:
1. some hatch fills msdfs being buggy or non connected 
2. text being too blurry when zooming in 

A. How to handle this function for "Hatch Box Fills"?
for hatch box fills we cannot use this `fwidth(uv)` because we're doing repeat sampling of a 8x8 msdf onto the screen! (that's ptobably you caused a double bug by the wrong division, two wrongs sometimes make a right :D)
so for hatch fills it's `float2 screenTexSize = float2(8,8)` or whatever the fixed size for them is!

B. How to handle this function for "glyph boxes"?
for glyph boxes (or I think you call the fontglyphs in other branch), the screenspace size of the glyph box can be computed in vshader and get passed along! no need for fwidth, and it will result in a nice shape with any rotation because the sdf would be exact as the last step (3rd one) above get's you correctly from uv distance to screenspace distance, then you can use that to compute opacity or whatever